src/adapters/vertexai/chat.ts

```diff
@@ -0,0 +1,208 @@
+    Object.assign(this, snapshot);
+  }
+}
+
+// Feel free to extend if you want to support additional parameters
+type VertexAIGenerateOptions = GenerateOptions;
+
+export interface VertexAIChatLLMInput {
+  modelId: string;
+  project: string;
+  location: string;
+  executionOptions?: ExecutionOptions;
+  cache?: LLMCache<VertexAIChatLLMOutput>;
+  parameters?: Record<string, any>;
+}
+
+export class VertexAIChatLLM extends ChatLLM<VertexAIChatLLMOutput> {
+  public readonly emitter: Emitter<GenerateCallbacks> = Emitter.root.child({
+    namespace: ["vertexai", "llm"],
+    creator: this,
+  });
+
+  protected client: GenerativeModel;
+
+  constructor(protected readonly input: VertexAIChatLLMInput) {
+    super(input.modelId, input.executionOptions, input.cache);
+    const vertexAI = new VertexAI({ project: input.project, location: input.location });
+    this.client = vertexAI.getGenerativeModel({
+      model: input.modelId,
+    });
+  }
+
+  static {
+    this.register();
+    Serializer.register(GenerativeModel, {
+      toPlain: (value) => ({
+        project: getPropStrict(value, "project"),
+        location: getPropStrict(value, "location"),
+        model: getPropStrict(value, "model"),
+        safetySettings: getPropStrict(value, "safetySettings"),
+        generationConfig: getPropStrict(value, "generationConfig"),
+        // systemInstruction: getPropStrict(value, "systemInstruction"),
+      }),
+      fromPlain: (value) => {
+        const vertexAI = new VertexAI({ project: value.project, location: value.location });
+        return vertexAI.getGenerativeModel({
+          model: value.model,
+          safetySettings: value.safetySettings,
+          generationConfig: value.generationConfig,
+        });
+      },
+    });
+  }
+
+  async meta(): Promise<LLMMeta> {
+    return { tokenLimit: Infinity };
+  }
+
+  async tokenize(input: BaseMessage[]): Promise<BaseLLMTokenizeOutput> {
+    const response = await this.client.countTokens({
+      contents: [{ parts: input.map((msg) => ({ text: msg.text })), role: "user" }],
+    });
+    return {
+      tokensCount: response.totalTokens,
+    };
+  }
+
+  protected async _generate(
+    input: BaseMessage[],
+    options: VertexAIGenerateOptions,
+    run: GetRunContext<this>,
+  ): Promise<VertexAIChatLLMOutput> {
+    // this method should do non-stream request to the API
+    // TIP: access inference parameters via `this.input.parameters` and `options`
+    // TIP: use signal from run.signal
+    const response = await signalRace(
+      () =>
+        this.client.generateContent({
+          contents: [{ parts: input.map((msg) => ({ text: msg.text })), role: "user" }],
+        }),
+      run.signal,
+      () => {},
+    );
+    let text = "Empty";
+    if (
+      response.response.candidates != undefined &&
+      response.response.candidates[0].content.parts[0].text != undefined
+    ) {
+      text = response.response.candidates[0].content.parts[0].text;
+    }
```

Comment:
```
const text = response?.response?.candidates?.[0]?.content?.parts?.[0]?.text || "Empty"
```

Also, I don't think we want to pass `Empty` (custom value). We should either throw an error or pass an empty string.

Reply:
What if there are any other `candidates`? Can this happen?

Reply:
I haven't seen it but I changed concatenating all candidates.